build_tools/bazel/iree_trace_runner_test.bzl

```
@@ -6,7 +6,222 @@
+        ] + runner_args,
+        data = [
+            ":%s" % bytecode_module_name,
+            ":%s" % trace,
+        ],
+        test_binary = trace_runner,
+        tags = tags + ["driver=%s" % driver],
+        timeout = timeout,
+        **kwargs
+    )
+
+def iree_single_backend_generated_trace_runner_test(
+        name,
+        generator,
+        trace_runner,
+        target_backend,
+        driver,
+        generator_args = [],
+        compiler_flags = [],
+        runner_args = [],
+        opt_tool = "//iree/tools:iree-opt",
+        opt_flags = [],
+        tags = [],
+        timeout = None,
+        **kwargs):
+    """Generates an iree_trace_runner_test using a custom python generator script.
+
+    The generator script produces a .mlir source and a .yaml trace, which are
+    passed to iree_trace_runner_test.
+
+    Args:
+        name: Name of the target
+        generator: Target to run to generate the source file and trace files. It will be
+            invoked with the following standard flags, in addition to GENERATOR_ARGS:
+            --output_code=(current binary dir)/name.mlir
+            --output_trace=(current binary dir)/name.yaml
+            --module_path=(current binary dir)/name.vmfb
+        generator_args: additional args to pass to the generator program.
+        target_backend: target backend to compile for.
+        driver: driver to run the module with.
+        compiler_flags: additional flags to pass to the compiler. Bytecode
+            translation and backend flags are passed automatically.
+        runner_args: additional args to pass to the trace-runner program. The driver
+            and input file flags are passed automatically.
+        tags: Additional labels to apply to the test. The package path and
+            "driver=${DRIVER}" are added automatically.
+        opt_tool: Defaulting to iree-opt. Tool used to preprocess the source files
+            if opt_flags is specified.
+        opt_flags: If specified, source files are preprocessed with opt_tool with
+            these flags.
+        trace_runner: trace-runner program to run.
+        timeout: timeout for the generated tests.
+        **kwargs: any additional attributes to pass to the underlying tests and test suite.
+    """
+
+    src = "%s.mlir" % (name)
+    trace = "%s.yaml" % (name)
+    module = "%s.vmfb" % (name)
+    native.genrule(
+        name = "%s_generate" % (name),
+        outs = [src, trace],
+        cmd = " ".join([
+            "$(location %s)" % (generator),
+            " ".join([('"%s"' % arg) for arg in generator_args]),
+            "--output_code=$(location %s)" % (src),
+            "--output_trace=$(location %s)" % (trace),
+            # Explanation for why "$(RULEDIR)/%s" instead of "$(location %s)" below:
+            # module_path points to a file that does not yet exist as it will
+            # be generated by iree_bytecode_module below iree_trace_runner_test.
```

```suggestion
            # be generated by iree_bytecode_module below by iree_trace_runner_test.
```

But also, can we just do this the other way around? Does the rule actually need it to exist ahead of time?

---

Unfortunately we can't, because the python script is co-generating the .mlir source with the .yaml trace.   We need that .mlir before we can pass it as the source to `iree_bytecode_module`, which will create the .vmfb which the trace is referencing...!

---

My take: maybe the .yaml trace shouldn't be referencing the .vfmb.... maybe it could be just describing how an interface is used without specifying an implementation of that interface. Then we would not have this issue here, and trace-runner programs would take the .vmfb as a separate commandline argument. @benvanik